File: 백준/15686_치킨배달/Main.java
import java.nio.file.Path;
import java.nio.file.Paths;
import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

class Solution {

    static class Pos {
        int r, c;

        Pos(int r, int c) {
            this.r = r;
            this.c = c;
        }

        Pos add(Pos other) {
            return new Pos(r + other.r, c + other.c);
        }

        @Override
        public boolean equals(Object obj) {
            if (obj instanceof Pos) {
                Pos other = (Pos) obj;
                return r == other.r && c == other.c;
            }
            return false;
        }

        @Override
        public int hashCode() {
            return Objects.hash(r, c);
        }
    }

    private int N, M;
    List<Pos> chickens;
    List<Pos> houses;
    private int[][] dists;

    public Solution(BufferedReader reader) throws IOException {
        String[] tokens = reader.readLine().split(" ");
        N = Integer.parseInt(tokens[0]);
        M = Integer.parseInt(tokens[1]);
        chickens = new ArrayList<>();
        houses = new ArrayList<>();


        for (int i = 0; i < N; i++) {
            String[] line = reader.readLine().split(" ");

            for (int j = 0; j < N ; j++) {
                if (line[j].equals("2")) {
                    chickens.add(new Pos(i, j));
                } else if(line[j].equals("1")) {
                    houses.add(new Pos(i, j));
                }
            }
        }

        dists = new int[houses.size() + 1][chickens.size() + 1];
    }

    private int calc(int selected) {
        int sum = 0;

        for(int i = 0 ; i < houses.size() ; i++) {
            int minValue = Integer.MAX_VALUE;
            for(int j = 0 ; j < chickens.size() ; j++) {
                if((selected & (0x01 << j)) == 0) {
                    continue;
                }
                minValue = Math.min(minValue, dists[i][j]);
            }
            if(minValue != Integer.MAX_VALUE) {
                sum += minValue;
            }
        }
        return sum;
    }

    private int combination(int selected, int idx, int count) {
        if(count == M) {
            return calc(selected);
        }

        int minValue = Integer.MAX_VALUE;

        for(int i = idx ; i < chickens.size() ; i++) {
            if((selected & (0x01 << i)) != 0) {
                continue;
            }

            int value = combination(selected | (0x01 << i), i, count + 1);
            minValue = Math.min(minValue, value);
        }

        return minValue;
    }

    public void run () {
        int answer = Integer.MAX_VALUE;

        for(int i = 0 ; i < chickens.size() ; i++) {
            Pos chicken = chickens.get(i);
            for(int j = 0 ; j < houses.size() ; j++) {
                Pos house = houses.get(j);
                dists[j][i] = Math.abs(chicken.r - house.r) + Math.abs(chicken.c - house.c);
            }
        }

        answer = combination(0, 0, 0);

        StringBuilder sb = new StringBuilder();
        sb.append(answer).append("\n");
        System.out.println(sb.toString());
    }
}

File: 백준/15686_치킨배달/main.rs
trait Solution {
    fn new()->Self ;
    fn run(&mut self);
}

struct Pos {
    r: i32,
    c: i32
}

struct Problem {
    n: usize,
    m: usize,
    chickens: Vec<Pos>,
    houses: Vec<Pos>,
    dists: Vec<Vec<i32>>
}

fn get_dist(a: &Pos, b: &Pos) -> i32 {
    (b.r-a.r).abs() + (b.c - a.c).abs()
}

impl Solution for Problem {

    fn new() -> Self {
        let tokens: Vec<i32> = IO::read_line()
            .split_whitespace()
            .map(|x| x.parse().unwrap())
            .collect();

        let n = tokens[0] as usize;
        let m = tokens[1] as usize;
        let mut chickens: Vec<Pos> = Vec::new();
        let mut houses: Vec<Pos> = Vec::new();

        for i in 0..n {
            let row: Vec<i32>= IO::read_line()
                .split_whitespace()
                .map(|x| x.parse().unwrap())
                .collect();

            for j in 0..n {
                if row[j] == 1 {
                    houses.push(Pos {r: i as i32, c: j as i32} );
                } else if row[j] == 2  {
                    chickens.push(Pos {r: i as i32 , c: j as i32} );
                }
            }
        }

        let mut dists: Vec<Vec<i32>> = vec![vec![0; chickens.len()]; houses.len()];

        for i in 0..houses.len() {
            for j in 0..chickens.len() {
                dists[i][j] = get_dist(&houses[i], &chickens[j]);
            }
        }


        Self {
            n: n,
            m: m,
            chickens: chickens,
            houses: houses,
            dists: dists
        }
    }

    fn run(&mut self) {
        let mut answer = self.combination(0, 0, 0);
        println!("{}", answer);
    }
}

impl Problem {

    fn simulate(&self, selected: i32) -> i32 {
        let mut sum: i32 = 0;

        for i in 0..self.houses.len() {
            let mut min_value = i32::MAX;
            for j in 0..self.chickens.len() {
                if selected & (0x01 << j) == 0 {
                    continue;
                }
                min_value = std::cmp::min(min_value, self.dists[i][j]);
            }

            if min_value != i32::MAX {
                sum += min_value;
            }
        }
        sum
    }

    fn combination(&mut self, idx: i32, count: i32, selected: i32) -> i32 {
        if count == self.m as i32 {
            return self.simulate(selected)
        }

        let mut min_value = i32::MAX;

        for i in idx .. self.chickens.len() as i32 {
            if (selected & (0x01 << i)) != 0 {
                continue;
            }
            let value = self.combination(i as i32, count + 1, selected | (0x01) << i);
            min_value = std::cmp::min(min_value, value);
        }

        min_value
    }
}

fn main() {
    let mut problem: Problem = Problem::new();
    problem.run();
}

mod IO {
    pub fn read_line() -> String {
        let mut buf = String::new();
        std::io::stdin().read_line(&mut buf).unwrap();
        buf.trim().to_string()
    }

    pub fn read_copy_line(buf: &mut String) {
        std::io::stdin().read_line(buf).unwrap();
    }
}

